### DISCUSSION TITLE
Is it possible to be PID controlled if temperature monitoring and SSR control are on different controllers?

### DISCUSSION THREAD

**Discussion by TheProgen** [artisan-roaster-scope/artisan#1243] [Q&A] [answered]
Hi everyone, first post here.

So I have this situation where my TC4+ started acting funny and I came across a sketch for an ESP32 board and MAX6675s emulating a TC4. I managed to get that running but I'm controlling the SSR with an analogue pot via the Arduino. Not a coder at all and it scared me to see so many dependencies in the aArtisan sketch so I didn't want to edit the code in any way and just run everything off the ESP32.

I've tried setting PID on in Control and the mode to Background but I don't see it taking over the SSR.

Is there anything else I should do?

**Comment by roasterdave** [comment]
The short answer to the question in your title is yes.  The process value (temperature) can come from a device different than where the control output (SSR) goes.  

From there your situation gets more complicated.  aArtisan was not authored by the Artisan team.  Maybe there are some community members here with knowledge that can help.  If not, you might want to try asking on the [Home Roasters forum](https://homeroasters.org/forum/index.php).  There is a lot of TC4 and related expertise over there.

**Reply by TheProgen** [reply, marked as answer]
I found out the answer. While you can have two devices as in my case, you cannot have PID control of the heater or fan if the Port is set to the microcontroller that's reading temperature.

I switched the port to the Arduino's and now I can see heater duty percentages.

However, I still had no PID control of the heater because now I don't have temperature readings going to Artisan. Guess I'll have to try taking a look at whether the short TC4 emulator code can be inserted into the aArtisan sketch and the MAX6675s go to the Arduino instead of ESP32 if I want PID control.

aArtisan covers a lot of devices so there's quite a bit of redundant code which can be removed. Good luck to me because I'm no coder. 😂

**Comment by darrenaddy** [comment]
I appreciate this question and answer. I was not previously aware of
aArtisan. There is this page on the subject on artisan-scope.org.

https://artisan-scope.org/devices/arduino/

Interestingly, it was updated just a week or so ago. The page contains this
warning:

*Watch out!* Older TC4 firmware versions operated at the lower 19200
baudrate.

Hope this helps someone.

Darren

On Thursday, July 27, 2023, Dave Baxter ***@***.***> wrote:

> The short answer to the question in your title is yes. The process value
> (temperature) can come from a device different than where the control
> output (SSR) goes.
>
> From there your situation gets more complicated. aArtisan was not authored
> by the Artisan team. Maybe there are some community members here with
> knowledge that can help. If not, you might want to try asking on the Home
> Roasters forum <https://homeroasters.org/forum/index.php>. There is a lot
> of TC4 and related expertise over there.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/artisan-roaster-scope/artisan/discussions/1243#discussioncomment-6570360>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AO5YMLU7LCT55HOWEKYGNPTXSL2UNANCNFSM6AAAAAA22JS5YM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: <artisan-roaster-scope/artisan/repo-discussions/1243/comments/6570360@
> github.com>
>


--

**Reply by MAKOMO** [reply]
Actually that page was update the last time about 9 month ago as can be seen from its repository. As this page does not define its own updated tag it inherits the update date from the whole site. This is misleading.

**Reply by TheProgen** [reply]
Marko, when we set a port in Artisan, is that for incoming only or both incoming and outgoing? 

My system has the port set to the virtual Bluetooth port which communicates with the ESP32 board. It works fine that way for reading temperatures but does Artisan send SSR duty signals via that port too?

**Reply by MAKOMO** [reply]
Serial ports are always used two way. How would otherwise Artisan be able to send out a request for data to the connected device in a request-response protocol?

**Reply by TheProgen** [reply]
It didn't matter at first because I was using the Arduino just to send bursts to the SSR. 😂

**Reply by TheProgen** [reply]
By the way, I came across an old post of yours. Will this approach still work with the current version of Artisan Scope?

![Screenshot_20230801-014440_Via_1](https://github.com/artisan-roaster-scope/artisan/assets/140711516/b53f4ebe-1397-47c1-96bb-c6e23da056f2)

**Reply by MAKOMO** [reply]
yes

**Reply by TheProgen** [reply]
Thanks! That saves a lot of trouble. I was looking around for a ESP32 based sketch which would integrate with Artisan for both temperature and PID control at first.

**Reply by TheProgen** [reply]
I tried the Dummy approach but I guess my situation is different from the person you were advising in that other forum because both my ESP32 and Arduino are emulating TC4's. I tried shifting ET and BT to TC channels 3 and 4 and then the Port to the Arduino's port but didn't know how to configure it to listen to the incoming temperature information from the ESP32. 😢

**Comment by darrenaddy** [comment]
Thanks for that clarification, Marko!

On Sunday, July 30, 2023, Marko Luther ***@***.***> wrote:

> Actually that page was update the last time about 9 month ago as can be
> seen from its repository. As this page does not define its own updated tag
> it inherits the update date from the whole site. This is misleading.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/artisan-roaster-scope/artisan/discussions/1243#discussioncomment-6587737>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AO5YMLVDZ35ZW4ZBUMJVVE3XSZ2C7ANCNFSM6AAAAAA22JS5YM>
> .
> You are receiving this because you commented.Message ID:
> <artisan-roaster-scope/artisan/repo-discussions/1243/comments/6587737@
> github.com>
>


--
